core/service/app.go

```diff
@@ -34,6 +44,272 @@ func (a App) FindAppInstallationsByTeamId(ct context.Context, teamID uint64) ([]
+		query := authorization.NewCreateAppVersionQuery(userID, appID)
+		hasPermission, err := a.authorizer.hasPermission(ct, query)
+		if err != nil {
+			a.authorizer.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: err})
+			return entity.AppVersion{}, err
+		}
+
+		if !hasPermission {
+			internalErr := &errs.Error{
+				Code:    errs.PermissionDenied,
+				Message: fmt.Sprintf("authorization query: %v", query),
+			}
+			a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: internalErr})
+			return entity.AppVersion{}, internalErr
+		}
+	}
+
+	av := entity.AppVersion{
+		AppID:          appID,
+		HasUIExtension: false,
+		IsPublic:       false,
+		CreatedAt:      time.Now().UTC(),
+	}
+	versionNum, err := a.appVersionDao.FindMaxVersionNumber(ct, appID)
+	if err != nil {
+		a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: err})
+		return entity.AppVersion{}, err
+	}
+	av.VersionNumber = versionNum + 1
+
+	err = a.appVersionDao.CreateAppVersion(ct, av)
+	if err != nil {
+		a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: err})
+		return entity.AppVersion{}, err
+	}
+
+	return av, nil
+}
+
+func (a App) UpdateAppVersion(ct context.Context, appID uint64, versionNumber int32, input UpdateAppVersionInput) (entity.AppVersion, *errs.Error) {
+	if feature.EnableAuthorization {
+		userID, ok := ctx.UserIDFromContext(ct)
+		if !ok {
+			internalErr := &errs.Error{
+				Code:    errs.NotFound,
+				Message: "user ID not found",
+			}
+			a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: internalErr})
+			return entity.AppVersion{}, internalErr
+		}
+
+		query := authorization.NewUpdateAppVersionQuery(userID, appID)
+		hasPermission, err := a.authorizer.hasPermission(ct, query)
+		if err != nil {
+			a.authorizer.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: err})
+			return entity.AppVersion{}, err
+		}
+
+		if !hasPermission {
+			internalErr := &errs.Error{
+				Code:    errs.PermissionDenied,
+				Message: fmt.Sprintf("authorization query: %v", query),
+			}
+			a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: internalErr})
+			return entity.AppVersion{}, internalErr
+		}
+	}
+
+	av, err := a.appVersionDao.FindAppVersionByAppIDAndVersionNumber(ct, appID, versionNumber)
+	if err != nil {
+		a.dataCollector.Logger.LogWithContext(ct, telemetry.Error, telemetry.Props{telemetry.CauseProp: err})
+		return entity.AppVersion{}, err
+	}
+
+	av.HasUIExtension = input.HasUIExtension
+	av.IsPublic = input.IsPublic
+	if input.IconURL != nil {
+		av.IconURL = input.IconURL
+	}
+	if input.Changes != nil {
+		av.Changes = input.Changes
+	}
+	if input.UIExtensionEntryPointPath != nil {
+		av.UIExtensionEntrypointPath = input.UIExtensionEntryPointPath
+	}
```

Comment:
```suggestion
	av.IconURL = input.IconURL
	av.Changes = input.Changes
	av.UIExtensionEntrypointPath = input.UIExtensionEntryPointPath
```

We do override to allow user to set the value to nil.